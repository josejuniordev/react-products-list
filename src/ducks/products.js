// ACTION TYPES
export const FETCH_PRODUCTS = 'FETCH_PRODUCTS';
export const FETCH_PRODUCTS_SUCCESS = 'FETCH_PRODUCTS_SUCCESS';
export const FETCH_PRODUCTS_FAILED = 'FETCH_PRODUCTS_FAILED';

// INITIAL STATE
export const INITIAL_STATE = {
  data: [],
  errors: {
    fetch: [],
  },
  loading: {
    fetch: false,
  }
};

// REDUCER
export default function(state = INITIAL_STATE, action) {
  const {type, products, errors} = action;

  switch (type) {
    case FETCH_PRODUCTS:
      return {
        ...state,
        loading: {
          ...state.loading,
          fetch: true,
        },
      };
    case FETCH_PRODUCTS_SUCCESS:
      return {
        ...state,
        data: products,
        loading: {
          ...state.loading,
          fetch: false,
        },
      };
    case FETCH_PRODUCTS_FAILED:
      return {
        ...state,
        loading: {
          ...state.loading,
          fetch: false,
        },
        errors: [...errors],
      };
    default:
      return state;
  }
};

// ACTION CREATORS
export function fetchProducts() {
  return {type: FETCH_PRODUCTS}
}

export function fetchProductsSuccess(products = []) {
  return {type: FETCH_PRODUCTS_SUCCESS, products}
}

export function fetchProductsFailed(errors = []) {
  return {type: FETCH_PRODUCTS_FAILED, errors}
}
