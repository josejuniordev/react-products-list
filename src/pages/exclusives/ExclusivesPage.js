import React, { Fragment, useEffect, useState } from 'react';
import ContentHeader from '../../components/layout/content-header/ContentHeader';
import { connect } from 'react-redux';
import ProductsList from '../../components/products-list/ProductsList';
import SearchBar from '../../components/search-bar/SearchBar';

export function ExclusivesPage(
  {
    products,
  }
) {
  const [productsData, setProductsData] = useState([]);
  const [exclusiveProducts, setExclusiveProducts] = useState([]);

  useEffect(() => {
    const filtered = filterProducts(products.data);
    setExclusiveProducts(filtered);
    setProductsData(filtered);
  }, [products]);

  function filterProducts(products = []) {
    return products.filter(product => product.exclusive);
  }

  function onSearchHandler(data) {
    setProductsData(data);
  }

  return (
    <Fragment>
      <ContentHeader
        title='Empresa XPTO'
        titleComplement='ConheÃ§a os produtos exclusivos'
        description='Listagem de produtos exclusivos - clique no produto desejado para saber mais'
        endEnhancer={<SearchBar data={exclusiveProducts} onSearch={onSearchHandler} />}
      />

      {
        products.loading.fetch
          ? <p>Buscando produtos...</p>
          : <ProductsList products={productsData} />
      }
    </Fragment>
  )
}

export default connect(({ products }) => ({products}))(ExclusivesPage);
