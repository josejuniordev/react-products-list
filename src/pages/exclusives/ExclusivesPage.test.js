import { shallow } from 'enzyme';
import React from 'react';
import {ExclusivesPage} from './ExclusivesPage';
import { INITIAL_STATE } from '../../ducks/products';

describe('<ProductPage />', () => {
  const rendered = shallow(<ExclusivesPage products={INITIAL_STATE}/>);

  it('should render correctly', () => {
    expect(rendered).toMatchSnapshot();
  });
});
