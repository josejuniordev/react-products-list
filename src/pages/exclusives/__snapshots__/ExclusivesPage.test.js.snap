// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductPage /> should render correctly 1`] = `ShallowWrapper {}`;
